orttraining/orttraining/python/training/ortmodule.py

```diff
@@ -76,6 +83,62 @@ def _onnx_value_info_to_buffer_tensor(value_info, device):
     dtype = _utils.dtype_onnx_to_torch(value_info.type.tensor_type.elem_type)
     return torch.zeros(shape, device=device, dtype=dtype)
 
+def _parse_inputs_for_onnx_export(module, inputs):
+    # Ignore optional *inputs explicitly specified as None
+    sig = signature(module.forward)
+    all_input_names = sig.parameters.keys()
+    input_names = []
+    dynamic_axes = {}
+    for input_idx, name in enumerate(all_input_names):
+        if input_idx < len(inputs) and inputs[input_idx] is not None:
+            input_names.append(name)
+            dynamic_axes[name] = {}
+            for dim_idx in range(len(inputs[input_idx].shape)):
+                dynamic_axes[name].update({dim_idx : 'input{}_dim{}'.format(input_idx, dim_idx)})
+    return input_names, dynamic_axes
+
+def _parse_outputs_for_onnx_export(module, inputs):
+    #   Do an inference to grab outputs
+    is_train_mode = module.training
+    module.eval()
+    with torch.no_grad():
+        # Deepcopy inputs, since input values may change after model run.
+        sample_inputs_copy = _deepcopy_model_input(*inputs)
+        try:
+            # Deepcopy model, in case model is stateful and changes after model run.
+            model_copy = copy.deepcopy(module)
+        except Exception:
+            model_copy = module
+            warnings.warn("This model cannot be deep copied (or pickled), which is a required step for stateful models to be properly exported to ONNX."
+                            " Compute will continue, but unexpected results may occur!")
+
+        sample_outputs = model_copy(*sample_inputs_copy)
+        def _create_output_dim_names(output, output_idx):
+            output_names, dynamic_axes = [], {}
+            name = 'out{}'.format(output_idx)
+            output_names.append(name)
+            dynamic_axes[name] = {}
+            for dim_idx in range(len(output.shape)):
+                dynamic_axes[name].update({dim_idx : '{}_dim{}'.format(name, dim_idx)})
+            return output_names, dynamic_axes
+
+        output_names = []
+        output_dynamic_axes = {}
+        if isinstance(sample_outputs, torch.Tensor):
+            output_names, output_dynamic_axes = _create_output_dim_names(sample_outputs, 0)
+        elif isinstance(sample_outputs, container_abcs.Mapping):
+            raise NotImplementedError('Dictionaries are not supported as output yet')
+        elif isinstance(sample_outputs, container_abcs.Sequence):
+            for idx, out in enumerate(sample_outputs):
+                tmp_output_names, tmp_output_dynamic_axes = _create_output_dim_names(out, idx)
```

Comment:
May we check here that the ```Sequence``` is a sequence of types we can work with (```Tensor```). If the user returns a sequence of custom type, that should be handled as well here, right?

Reply:
Will add logic to `_create_output_dim_names` to catch this